src/Providers/ProviderRow.tsx

```diff
@@ -0,0 +1,122 @@
+import React, { JSXElementConstructor } from 'react';
+import { Link } from 'react-router-dom';
+import { RowProps } from 'src/components/TableView';
+import { CONDITIONS, PROVIDERS, useTranslation } from 'src/internal/i18n';
+import {
+  ACTIONS,
+  HOST_COUNT,
+  NAME,
+  NAMESPACE,
+  NETWORK_COUNT,
+  READY,
+  STORAGE_COUNT,
+  TYPE,
+  URL,
+} from 'src/utils/constants';
+
+import { PATH_PREFIX } from '@app/common/constants';
+import { StatusIcon } from '@migtools/lib-ui';
+import { ResourceLink } from '@openshift-console/dynamic-plugin-sdk';
+import { Button, Popover } from '@patternfly/react-core';
+import { DatabaseIcon, NetworkIcon, OutlinedHddIcon } from '@patternfly/react-icons';
+import { Td, Tr } from '@patternfly/react-table';
+
+import { MergedProvider } from './data';
+import { ProviderActions } from './providerActions';
+
+interface CellProps {
+  value: string;
+  entity: MergedProvider;
+  t?: (k: string) => string;
+}
+const StatusCell = ({ value, entity: { conditions }, t }: CellProps) => {
+  const existingConditions = Object.values(conditions).filter(Boolean);
+  const toState = (value) => {
+    switch (value) {
+      case 'True':
+        return 'Ok';
+      case 'False':
+        return 'Error';
+      default:
+        return 'Unknown';
+    }
+  };
+  const label = CONDITIONS?.[value]?.(t);
+  return (
+    <Popover
+      hasAutoWidth
+      bodyContent={
+        <div>
+          {existingConditions.length > 0
+            ? existingConditions.map(({ message, status }) => {
+                return <StatusIcon key={message} status={toState(status)} label={message} />;
+              })
+            : t('No information')}
+        </div>
+      }
+    >
+      <Button variant="link" isInline aria-label={label}>
+        <StatusIcon status={toState(value)} label={label} />
+      </Button>
+    </Popover>
+  );
+};
+
+const TextCell = ({ value }: { value: string }) => <>{value ?? ''}</>;
+
+const TextWithIcon = ({ value, Icon }: { value: string; Icon: JSXElementConstructor<unknown> }) => (
+  <>
+    {value && (
+      <>
+        <Icon /> <TextCell value={value} />
+      </>
+    )}
+  </>
+);
+
+const ProviderLink = ({ value, entity }: CellProps) => (
+  <ResourceLink kind={entity.kind} name={value} namespace={entity?.namespace} />
+);
+
+const HostCell = ({ value, entity: { ready, name, type } }: CellProps) => (
+  <>
+    {ready === 'True' && value && type === 'vsphere' ? (
+      <Link to={`${PATH_PREFIX}/providers/vsphere/${name}`}>
+        <TextWithIcon Icon={OutlinedHddIcon} value={value} />
+      </Link>
+    ) : (
+      <TextWithIcon Icon={OutlinedHddIcon} value={value} />
+    )}
+  </>
+);
+
+const cellCreator = {
+  [NAME]: ProviderLink,
```

Comment:
I'm torn as to how this looks.  All the internals are just references to the functional components while the others are functional components.  The visual difference between the two is interesting.

What about just doing all of them the same way:
```jsx
  [NAME]: (cellProps) => <ProvierLink {...cellProps} />,
```

Reply:
IMHO the different approaches are justified:
1. References are used in cases where the component would be too long to fit into one line i.e. `ProviderLink` (and with TS typings it's getting harder to write a one liner:))
2. The anonymous components are used when simple customization is required i.e. different icon. Creating a named component is not needed here (cells are not public  - encapsulated inside `ProviderRow`)

Reply:
> IMHO the different approaches are justified:
> 
>    1. References are used in cases where the component would be too long to fit into one line i.e. `ProviderLink` (and with TS typings it's getting harder to write a one liner:))

With typing on `cellCreator` really not much else needs to be done.  The `{...cellProps}` works just fine w/o complaint from TS (well at least when I was tinkering with it before).  This way just makes it more obvious that ALL cellProps are going to the component.

> 
>    2. The anonymous components are used when simple customization is required i.e. different icon. Creating a named component is not needed here (cells are not public  - encapsulated inside `ProviderRow`)

That's a fair point.

In summary, leaving it how you have it is ok.  It's reasonable to have a think about other forms the code can take and if anything is to be gained by it.

Reply:
(continue over on https://github.com/kubev2v/forklift-console-plugin/pull/33#discussion_r1024387385)